ADD checksums so building storybook is faster

ndelangen opened the pull request:
Issue:

`yarn bootstrap --build` is dumb it regenerates the dist, even if the src hasn't changed at all

## What I did
I added a checksum mechanism

The packages will now only be "prepared" if the src folder has any changes.

ndelangen committed:
ADD checksums so building storybook is faster, it only builds if src has changed

yannbf commented on scripts/prepare.js:
```diff
@@ -1,18 +1,51 @@
+
+  const future = await createChecksum();
+
+  return existing === future ? existing : future;
```
Will this equality check work with comparing objects? Won't you need to use something like `isDeepStrictEqual`?

ndelangen replied:
it's not comparing objects it's comparing strings AFAIK.

yannbf replied:
I assumed it was an object because of this line `const { checksum: existing } = info.package.storybook || {};`

ndelangen replied:
info.package.storybook.checksum is a string, I call this string "existing"

yannbf commented on scripts/prepare.js:
```diff
@@ -28,48 +61,83 @@ const ignore = [
+run().catch((e) => {
+  console.error(e);
+  process.exitCode = 1;
+  // throw e;
```
is this comment necessary?

yannbf reviewed (commented):
This is a game changer 🙌

ndelangen committed:
IMPROVE build & bootstrap script

ndelangen committed:
ran yarn script silently

ndelangen committed:
IMPROVE existence checks

ndelangen committed:
change parallel to stream, because we need topological order

ndelangen committed:
IMPROVE error handling

ndelangen committed:
CLEANUP

ndelangen committed:
FIX cyclical dependency && FIX exists check

ndelangen committed:
Merge branch 'next' into tech/checksum-builds

ndelangen committed:
UPDATE checksums && disable yarn silent

ndelangen committed:
Merge branch 'next' into tech/checksum-builds

ndelangen committed:
FIX test

ndelangen committed:
Merge branch 'next' into tech/checksum-builds

ndelangen committed:
Merge branch 'next' into tech/checksum-builds

# Conflicts:
#	lib/ui/package.json
#	scripts/build-package.js

ndelangen committed:
UPDATE checksums && CHANGE build confirmation

ndelangen committed:
CLEANUP

ndelangen committed:
IMPROVE reset & prepare scripts

ndelangen commented:
I changed the location of the checksums (no longer in package.json) thanks for the suggestion @tooppaaa 

I changed the reset script so it no longer removes the dist folder

ndelangen committed:
Merge branch 'next' into tech/checksum-builds

# Conflicts:
#	lib/ui/package.json
#	package.json
#	scripts/build-package.js
#	scripts/prepare.js
#	scripts/utils/watch-babel.js
#	scripts/utils/watch-tsc.js

ndelangen committed:
MIGRATE reset script to TS && FIX lockfile

ndelangen committed:
Speed up the build with --incremental && --skip-initial-build

ndelangen committed:
MIGRATE prepare to typescript

ndelangen committed:
CLEANUP

ndelangen committed:
FIX

ndelangen committed:
this dependency is needed unfortunately

ndelangen committed:
remove silent for debugging

ndelangen committed:
use build-script for bootstrap

ndelangen committed:
Merge branch 'next' into tech/checksum-builds

ndelangen committed:
Merge branch 'next' into tech/checksum-builds

# Conflicts:
#	yarn.lock

ndelangen committed:
FIX lockfile

ndelangen committed:
ADD large resource to circleci

ndelangen committed:
ADD --transpile-only

ndelangen commented:
Thank you for the assist @KyleTryon !

It seem the bigger instance fixes it. I'm going to try a few optimizations

yannbf commented on scripts/run-reset.ts:
```diff
@@ -7,25 +7,25 @@ const logger = console;
+            uri.match(/.idea/) ||
+            uri.match(/reset.log/)
+          ) {
+            // ignore, don't remove these files
```
I think you can extract all these uri match assertions to variables and as a result you both have more readable code plus you can re-structure these conditionals to be simpler. This empty if is a bit weird

yannbf commented on scripts/utils/compile-babel.js:
```diff
@@ -30,37 +29,42 @@ function getCommand(watch) {
-  const { watch = false, silent = true, errorCallback } = options;
+async function babelify(options = {}) {
+  const { watch = false, silent = !watch } = options;
+  const [src] = await Promise.all([exists('src')]);
```
why not `const src = await exists('src');` ?

yannbf reviewed (changes requested):
Checked out and tried in my machine, works wonderfully!
However I'd HIGHLY recommend running this in a windows machine before merging. I can try that tomorrow and let you know!

My change request is simple, related to `build-package.js` please consider it and let me know if you need a hand 👍

ndelangen commented on lib/api/src/version.ts:
```diff
@@ -1 +1,11 @@
+// @ts-ignore
+import preval from 'preval.macro';
+
+export const version = preval`
```
This is executed at build-time, the resulting bundle will only contain a string. And it's all automatic.

ndelangen commented on scripts/prepare.ts:
```diff
@@ -0,0 +1,212 @@
+import * as shell from 'shelljs';
+import chalk from 'chalk';
+import fse from 'fs-extra';
+import { PackageJson } from 'type-fest';
```
Never makes it into a bundle we ship?

ndelangen commented on scripts/prepare.ts:
```diff
@@ -0,0 +1,212 @@
+    }
+
+    // Remove all copied TS files (but not the .d.ts)
+    if (/\.tsx?$/.test(filePath) && !/\.d\.ts$/.test(filePath)) {
```
it is?

ndelangen committed:
Merge branch 'next' into tech/checksum-builds

# Conflicts:
#	lib/api/src/version.ts
#	lib/ui/package.json

ndelangen committed:
WIP

ndelangen committed:
Remove esm and use ts-node everywhere for scripts

ndelangen committed:
more TS migration

ndelangen committed:
change to use ts-node-script

ndelangen committed:
Merge branch 'next' into tech/checksum-builds

# Conflicts:
#	addons/graphql/package.json
#	lib/api/src/version.ts
#	lib/components/package.json
#	lib/ui/package.json

ndelangen committed:
try using normal circle runner again

ndelangen commented on scripts/build-package.js:
```diff
@@ -126,44 +140,45 @@ function run() {
-          spawn(`lerna run prepare --scope "${glob}"`);
+          let confirmation = true;
+
+          if (packageNames.length > 5) {
```
not a bad merge at all. I found creating the globs is more error prone then just listing all packages in separate flags.

ndelangen committed:
CLEANUP

ndelangen committed:
big runner again

ndelangen committed:
CLEANUP

ndelangen committed:
fix downgrade script

ndelangen committed:
FIX

tooppaaa commented on lib/cli/package.json:
```diff
@@ -52,8 +53,8 @@
     "fs-extra": "^9.0.0",
     "get-port": "^5.1.1",
     "inquirer": "^7.0.0",
+    "json5": "^2.1.3",
```
Order was right before, final reference of this in Meteor generator. Worth removing it !

tooppaaa commented on scripts/prepare.ts:
```diff
@@ -0,0 +1,169 @@
+  const { hash } = await hashElement(path.join(modulePath, 'src'), {
+    folders: { exclude: ['tests', '__mocks__', '__tests__'] },
+    files: {
+      include: ['*.js', '*.jsx', '*.ts', '*.jsx', '*.json'],
```
**jsx is here twice. Missing tsx.**

Random question, is worth to create 2 hashes, one for src, one for ts, tsx which would only do tsc when this second hash changes.

It may add another level of complexity to this and maybe not needed !

tooppaaa reviewed (changes requested):
Only the hash include extensions to update.
Awesome otherwise !

ndelangen committed:
Merge branch 'next' into tech/checksum-builds

# Conflicts:
#	yarn.lock

ndelangen committed:
CHANGE so verdaccio is a normal dev dependency, since it's light anyway

ndelangen committed:
Merge branch 'next' into tech/checksum-builds

# Conflicts:
#	addons/graphql/package.json
#	lib/api/src/version.ts
#	lib/components/package.json
#	lib/ui/package.json

ndelangen committed:
Merge branch 'next' into tech/checksum-builds

ndelangen committed:
transpile only for ts-node for our scripts

ndelangen committed:
only generate downgrades when specifically asked for

ndelangen committed:
CLEANUP

ndelangen committed:
exclude stories from ts

ndelangen committed:
FIX lockfile

ndelangen committed:
CLEANUP

ndelangen committed:
FIX linting

ndelangen committed:
FIX cypress
